### ISSUE TITLE
use Java11 Features requires Unit Cache Cleanup to work

### ISSUE THREAD

**Issue by MCMicS** [gwtproject/gwt#9688]
**GWT version: 2.9.0** 
**Operating System: Windows 10, Linux** 
---

##### Description
Collected from Gitter (@eguardiola, @MCMicS, @niloc132) 
Reproduce it with Hello Sample

##### Steps to reproduce
1. `ant gwtc` for Hello sample
2. change in Hello.java:
```
public class Hello implements EntryPoint {

  public void onModuleLoad() {

    var b = new Button("Click me", (ClickHandler)e -> {
        Window.alert("Hello, AJAX");
    });

    RootPanel.get().add(b);
  }
}
```
3. change build.xml
```
 <target name="javac" depends="libs" description="Compile java source to bytecode">
    <mkdir dir="war/WEB-INF/classes"/>
    <javac srcdir="src" includes="**" encoding="utf-8"
        destdir="war/WEB-INF/classes"
        source="11" target="11" nowarn="true"
        debug="true" debuglevel="lines,vars,source">
      <classpath refid="project.class.path"/>
    </javac>
    <copy todir="war/WEB-INF/classes">
      <fileset dir="src" excludes="**/*.java"/>
    </copy>
  </target>

  <target name="gwtc" depends="javac" description="GWT compile to JavaScript (production mode)">
    <java failonerror="true" fork="true" classname="com.google.gwt.dev.Compiler" maxmemory="512m">
      <classpath>
        <pathelement location="src"/>
        <path refid="project.class.path"/>
      </classpath>
      <arg line="-war"/>
      <arg value="war"/>
      <!-- Additional arguments like -style PRETTY or -logLevel DEBUG -->
      <arg line="-sourceLevel"/>
      <arg value="1.11" /> <!-- or auto -->
      <arg line="${gwt.args}"/>
      <arg value="com.google.gwt.sample.hello.Hello"/>
    </java>
  </target>
```
4. Compile it again leads to following error

> [ERROR] Line 31: var cannot be resolved to a type

##### Known workarounds
* Delete the _gwt-unitCache_ and compile again will work
* use following if compile first with 1.7 level.
```
<arg line="-sourceLevel"/>
<arg value="auto" />
```
* `jvm arg -Dgwt.persistentunitcache=false` to disable cache. But compilation will be slower

**Comment by axls** [comment]
@MCMicS I saw this error when used wrong source level by mistake. And changing the source level to the correct value has not been helping until removing GWT unit cache. But after cache removing the compilation (including incremental too) should work. Could you confirm, that the issue are happening only first time when you change source level without removing cache?

**Comment by MCMicS** [comment]
Yes can I confirm. In gitter chat we discuss this too.

**Comment by rluble** [comment]
That should be easy to fix by adding the version to the gwt-unitCache directory name [here](
https://github.com/gwtproject/gwt/blob/178c60a224e9ff8b9e05904e099a91c0e12f107e/dev/core/src/com/google/gwt/dev/javac/PersistentUnitCacheDir.java#L52)

Incompatible compilations should have a different unitCache, since the unit cache only caches bytecode, I expect that just adding the target version to the name would fix the issue.

**Comment by MCMicS** [comment]
@rluble >This sounds ok. @niloc132 has following possible solution for this:
> a workaround within the compiler could be to purge the cache if the sourceLevel arg changes between compiles

I think to create cache by source/target version is the best option

**Comment by rluble** [comment]
Either solution is ok, but I don't think you have the source/target version directly stored in the cache (except as the class file version in every CompiledUnit, but it would be more work IMHO as you need to retrieve it and keep track of the relation between class file version and java version). 

Keeping separate caches is the simplest solution.

**Comment by MCMicS** [comment]
@rluble Do you mena it is better to change the cache directory or add level to Unit Cache creation?

UnitCacheSingleton
```
  public static String getRelevantOptionsHash(JJSOptions options) {
    return StringUtils.toHexString(
        Md5Utils.getMd5Digest(
            Joiner.on('-').join(
                options.getJsInteropExportFilter(),
                options.shouldGenerateJsInteropExports(),
                options.getSourceLevel().getStringValue()
        ).getBytes()));
  }
```

**Comment by niloc132** [comment]
That's the fix I've done locally, and it seems to pass my initial testing, I'll put up a review after my morning meetings.

**Comment by MCMicS** [comment]
@niloc132 My locally fix works too on Win10.

But this only works if `<arg line="-sourceLevel"/>` is set on gwtc target. If not and you only change the source and target from _javac_ target then it will still fail

But this is maybe a problem with auto JVM detection

**Comment by niloc132** [comment]
https://gwt-review.googlesource.com/c/gwt/+/22560 is a patch for this, @MCMicS perhaps you can try this again with java14 on windows, with some extra logging of the options value and the hash to confirm that the old unitCache is not being picked up? My hunch is that this is related to your other windows issues of the FS showing that cache entries were truncated...

--

That said, if the user has no `-sourceLevel` set and uses the same JVM >= 11 each time they build, wouldn't the `options.getSourceLevel()` be consistent anyway (i.e. "JAVA11"), so the cache name would be unchanged? 

In other words, if the default of "JAVA11" is "changed" to be explicitly "JAVA11", that isn't a change. The only real changes happening in this test case are a) `Button` -> `var`, and b) javac changing from source/target from "1.7" to "11". Why is this working (on non-win10 machines)?

**Comment by MCMicS** [comment]
@niloc132 
The patch work but the first detected source Level is Java8 instead Java11
```
// First
gwtc:
     [java] 1.8
     [java] C90BB3B791D3293DB8D5BA1AB0A974D4

// with sourceLevel auto (explicitly set)
gwtc:
     [java] 1.11
     [java] 18CAE75E829AF224BF18DEE97AC379EF
```

so it seems that default value for unset sourceLevel is Java8. See https://github.com/gwtproject/gwt/blob/f031aba1e5aff65143005de841a491b42fe1031a/dev/core/src/com/google/gwt/dev/util/arg/ArgHandlerSourceLevel.java#L35

The Filesystem still occurs if class files are level 11 and gwt sourceLevel is 1.8 (only first cache creation)

**Comment by niloc132** [comment]
You're right of course, thanks for pointing out my mistake - the SourceLevel.DEFAULT_SOURCE_LEVEL only kicks in if you don't get to the compiler via args, otherwise the arg handler takes over and decided that JAVA8 is the default.

**Comment by rluble** [comment]
> @rluble Do you mena it is better to change the cache directory or add level to Unit Cache creation?

`getRelevantOptionsHash()` is actually the right place.

**Comment by rluble** [comment]
> @rluble Do you mena it is better to change the cache directory or add level to Unit Cache creation?

`getRelevantOptionsHash()` is actually the right place.

**Comment by niloc132** [comment]
Patch merged, thanks for your help in quickly tracking this down everyone!
